pkg/kine/drivers/generic/batch.go

```diff
@@ -0,0 +1,206 @@
+package generic
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+
+	"github.com/canonical/k8s-dqlite/pkg/kine/prepared"
+)
+
+type BatchConn interface {
+	ExecContext(ctx context.Context, query string, args ...any) (result sql.Result, err error)
+}
+
+var _ BatchConn = &sql.DB{}
+var _ BatchConn = &sql.Tx{}
+var _ BatchConn = &sql.Conn{}
+
+var _ BatchConn = &prepared.DB{}
+var _ BatchConn = &prepared.Tx{}
+
+type batchStatus int
+
+const (
+	batchNotStarted batchStatus = iota
+	batchStarted
+	batchRunning
+)
+
+type Batch struct {
+	db     *prepared.DB
+	mu     sync.Mutex
+	cv     sync.Cond
+	status batchStatus
+
+	queue []*batchJob
+	runId int64
+}
+
+func NewBatch(db *prepared.DB) *Batch {
+	b := &Batch{
+		db: db,
+	}
+	b.cv.L = &b.mu
+	return b
+}
+
+func (b *Batch) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	runId := b.runId
+	if b.status == batchRunning {
+		// The current run is already taking place.
+		runId++
+	}
+
+	job := &batchJob{
+		ctx:   ctx,
+		query: query,
+		args:  args,
+		runId: runId,
+	}
+	b.queue = append(b.queue, job)
+
+	b.run()
+
+	for job.runId >= b.runId {
+		b.cv.Wait()
+	}
+
+	if job.err != nil {
+		return nil, job.err
+	}
+
+	return job, nil
+}
+
+// run starts a batching job which will run until queue exaustion.
+// run does not block other goroutine from enqueuing new jobs.
+//
+// It must be called while holding the mu lock.
+func (b *Batch) run() {
```

Comment:
Do we execute b.run() on every call on execute? Would it make sense to wait until we have a couple of txn or a timer runs out to trigger a batch?

Reply:
This is executed on every call, but only some will actually spawn a goroutine. The goroutine will stay alive until the queue is empty (which might be for a long time, depending on the status of the queue).

Regarding adding a slight timer to that, it might indeed help. However this is the kind of stuff that I prefer not to do as:
 - we would introduce more latency
 - we don't know how much time to wait (I suspect that the optimal one would be different depending on disk performance)
 
 This one has the first transaction "likely" with a single statement, but then all the other of the burst should be grouped by a time which is the commit time which should decently mimic the timer you are talking about.